build: Optionally use git for all docker builds with BUILDKIT

jrajahalme opened the pull request:
Make building with DOCKER_BUILDKIT and using a shallow git clone as
the build context opt-in: the developer must define DOCKER_BUILDKIT to
use these. When DOCKER_BUILDKIT is defined, the following takes place:

1. A separate docker build directory for all docker image targets is
used. This makes sure the build context only contains the files in git
and none of the other files that may exist in the working
directory. '_build' in the repo is used as the root of the build
directory by default (can be overridden by defining BUILD_DIR). A
shallow bare git clone of the main repo is created in '_build/.git',
which is checked out into '_build/context'.

The build context is subsequently synced with git fetch to avoid
overwriting all files. This makes docker build context sync much
faster and allows docker to use cached build stages.

2. The .git directory is left out of the build context, so that any
changes in there will not affect docker build stage caching. During
the build context sync a file '_build/context/.git' points to
'_build_.git', but that is removed before the context is handed to
docker.

3. BUILD_DIR can be passed in to specify an alternate build root
(replacing the default '_build'). This is beneficial in keeping the
main repo directory cleaner, and also to avoid slow file I/O in a
shared file system (e.g, between the host and a CI VM).

A new make target `veryclean` that guts docker caches has been added.
This may help resolve the issue where Docker fails to actually load an image
(e.g., cilium-builder) which leads to weird issues like `/bin/sh` not found.

Testing this on a MacBook Pro yielded the following results:

- Clean and build cilium, cilium-docker-plugin, cilium-operator, and hubble-relay

Before these changes:
$ time make docker-image
8m17.558s

After:
$ time make docker-image
5m35.515s

$ DOCKER_BUILDKIT=1 time make docker-image
2m49.41s

Fixes: #11443

aanm commented on hubble-relay.Dockerfile:
```diff
@@ -1,6 +1,9 @@
+#
```
I have to ask, why the `#` on top of all files?

jrajahalme replied:
It makes it a bit easier to add the required stanza to the first line without risk of doing it if the Dockerfile already contains it.

jrajahalme committed:
contrib: Update check scripts

Add exclusion of _build to pass checks, as otherwise
'_build/context/vendor/' would cause problems. Exclude also '.git' to
speed them up. Unify syntax to use grep's
--exclude-dir={dir1,dir2,...} syntax.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
.gitignore: Add 'test/*tmp'

Add 'test/*tmp' to cover 'test/gke@tmp/' and anything like it in future.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
build: Avoid using git if not in a git repo

Do not use git if not in a git repo.

Only create GIT_VERSION if the existing file is already not the
same. This helps docker caching.

Store the list bpf files to a temporary file BPF_SRCFILES, which is
ignored by git like GIT_VERSION. This allows builds to succeed without
git.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
build: Optionally use git for all docker builds with BUILDKIT

Make building with DOCKER_BUILDKIT and using a shallow git clone as
the build context opt-in: the developer must define DOCKER_BUILDKIT to
use these. When DOCKER_BUILDKIT is defined, the following takes place:

1. A separate docker build directory for all docker image targets is
used. This makes sure the build context only contains the files in git
and none of the other files that may exist in the working
directory. '_build' in the repo is used as the root of the build
directory by default (can be overridden by defining BUILD_DIR). A
shallow bare git clone of the main repo is created in '_build/.git',
which is checked out into '_build/context'.

The build context is subsequently synced with git fetch to avoid
overwriting all files. This makes docker build context sync much
faster and allows docker to use cached build stages.

2. The .git directory is left out of the build context, so that any
changes in there will not affect docker build stage caching. During
the build context sync a file '_build/context/.git' points to
'_build_.git', but that is removed before the context is handed to
docker.

3. BUILD_DIR can be passed in to specify an alternate build root
(replacing the default '_build'). This is beneficial in keeping the
main repo directory cleaner, and also to avoid slow file I/O in a
shared file system (e.g, between the host and a CI VM).

Testing this on a MacBook Pro yielded the following results:

- Clean and build cilium, cilium-docker-plugin, cilium-operator, and hubble-relay

Before these changes:
$ time make docker-image
8m17.558s

After:
$ time make docker-image
5m35.515s

$ DOCKER_BUILDKIT=1 time make docker-image
2m49.41s

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

errordeveloper commented on Makefile:
```diff
@@ -282,16 +290,82 @@ docker-cilium-image-for-developers:
+ifdef DOCKER_BUILDKIT
+	# Export with value expected by docker
+	export DOCKER_BUILDKIT=1
+	BUILDKIT_DOCKERFILE_FILTER := | sed -e "1s|^\#.*|\# syntax = docker/dockerfile:experimental|" -e "s|^RUN\(.*\)make|RUN --mount=type=cache,target=/root/.cache/go-build\1make|"
```
I would much rather see this added permanently, what earliest version of Docker supports this option?

errordeveloper replied:
I spoke with @nebril and it sounds like upgrading Docker in Jenkins wouldn't be a big deal :)

jrajahalme replied:
dev VM has docker 19.03 and it still has a bug where it sometimes fails to pull the base image on the FROM line, behaving as if it was pulled, but then "make" or "/bin/sh" can't be found (see @joestringer's comments below). At the least, enabling DOCKER_BUILDKIT in the CI takes some testing in addition updates (if needed).

jrajahalme committed:
build: Fix parallel build dependency on clean

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
build: Add 'veryclean' make target to gut docker

Add new 'veryclean' target that cleans up the docker (desktop)
instance, useful if docker's file system becomes full.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on Makefile:
```diff
@@ -250,11 +262,11 @@ clean-container:
+veryclean:
+	-$(QUIET) $(CONTAINER_ENGINE) image prune -af
+	-$(QUIET) $(CONTAINER_ENGINE) builder prune -af
+	-$(QUIET) $(CONTAINER_ENGINE) system prune -af
```
:exploding_head:

From https://docs.docker.com/engine/reference/commandline/system_prune/ :

> $ docker system prune -a --volumes
> 
> WARNING! This will remove:
>         - all stopped containers
>         - all networks not used by at least one container
>         - all volumes not used by at least one container
>         - all images without at least one container associated to them
>         - all build cache
> Are you sure you want to continue? [y/N] y

jrajahalme committed:
build: Set buildkit git origin to a relative path

Set the path of the git origin for the build context shallow clone to
a relativa path so that it also works when build directory was created
in a mounted directory (host vs. VM).

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
